### ISSUE TITLE
[REQ] Don't switch Peloton classes after a ride has started when a new class detected

### ISSUE THREAD

**Issue by Brodie01** [cagnulein/qdomyos-zwift#2098]
**Is your feature request related to a problem? Please describe.**
Peloton subscriptions seem to allow you to do concurrent classes on different devices using the same account. My wife will sometimes do a yoga class in Peloton when I'm doing a ride. IF I start my ride first and sync to Peloton, when she starts her Yoga class, QZ will see the class start and will pop up asking me to sync with it. If I select no, it still starts tracking the yoga class anyway. 

**Describe the solution you'd like**
Ideally, if I start a ride in QZ and sync to a peloton class, QZ will stop monitoring peloton for new classes that start. OR if it is useful to be able to have QZ switch, perhaps have a pop up saying that QZ has detected a new class, do you want to leave current class or stay.

**Describe alternatives you've considered**
No other solutions I can think of short of not working out at the same time or Perhaps it may work if she starts her yoga class first - but that's not always a discussion we have.

**Additional context**

**Comment by cagnulein** [comment]
@Brodie01 Hi, if you press no, qz will still use the current workout unless I'm not thinking something. Could you please share a debug log when you are doing this?

Unfortunately in Italy, peloton is not allowed so I can't test this myself 

Let me know

**Comment by cagnulein** [comment]
This way https://github.com/cagnulein/qdomyos-zwift/wiki/How-do-i-get-the-debug-log-in-case-something-doesn't-work%3F

**Comment by cagnulein** [comment]
If you don't have it I can check the code and try to understand it

**Comment by marklogan17** [comment]
Hey @cagnulein - happened to me today with about 5 or 6 minutes left in my ride.  I hit 'yes' because I wasn't really thinking nor expecting this to happen - shared my account with someone a few weeks ago and this is the first time we ended up using it at the same time.  Anyway, the stats worked OK - average power, etc didn't get lost but of course the class was messed up.  Not sure if Peloton API will provide access to both streams at the same time or only one.  Happy to test it for you.  I'll grab the log file and upload it here shortly from today's so you can at least see what's going on.

**Comment by marklogan17** [comment]
Here's the log.  Around 6 minutes left you'll see the second class come in.  Honestly I think if you're in the middle of a class I just wouldn't even prompt at all.  Ignore a new class and just stay with the current one.

One other thing you will find in this log - I've been meaning to ask you but I keep forgetting...  When I hit 'yes' to sync the class the resistance drops WAY down to like 10 (Echelon resistance) so I have to set the gears to like +15 to get to the level I want.  I have min resistance set to 20 but it still goes down to 10 or so at the start.  Once I get it where I want it I'm fine but the Echelon vs Peloton conversion stays way off.  A Peloton 45 resistance with +15 gears is equating to a 27 Echelon resistance.  I'm assuming I've messed with some setting or there's something about the SmartSpin that's doing it.  Not a big deal at all (i.e. don't waste any time) but if there's something obvious let me know. 

[debug-Mon_Feb_12_07_22_21_2024.zip](https://github.com/cagnulein/qdomyos-zwift/files/14241123/debug-Mon_Feb_12_07_22_21_2024.zip)

**Comment by cagnulein** [comment]
Ok, I'll check it out. thanks!


Il giorno lun 12 feb 2024 alle 17:33 marklogan17 ***@***.***>
ha scritto:

**Comment by cagnulein** [comment]
i checked but there is something strange for the main issue. I see this

Mon. Feb. 12 07:23:13 2024 1707740593464 Debug: ../src/homeform.cpp void
homeform::peloton_start_workout() "peloton_start_workout!"
Mon. Feb. 12 07:46:31 2024 1707741991279 Debug: ../src/homeform.cpp void
homeform::peloton_start_workout() "peloton_start_workout!"

so this means that you press yes on the popup both times, is it right? if
so QZ is doing the right thing, discarding the current workout to the next
one, am I missing something?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 12 feb 2024 alle ore 17:45 cagnulein ***@***.***> ha
scritto:

**Comment by marklogan17** [comment]
In this case I did hit Yes the second time.  I think Jay was saying that he hit 'no' and it restarted from the beginning - I don't have a log for that but I can try it and send one.  I think the thing is that if you're in a ride maybe you want 'yes' to swap, 'no' to basically stop syncing and 'continue current' to just keep going - or alternatively, ignore the new class and just keep going without prompting at all.

I forgot to say for the second thing I mentioned - in the log for today I had forgotten to start QZ (it was paused) so the resistance drop was when I hit start as opposed to when I hit 'yes'.  Normally I have hit start already and so the resistance drop is when I hit 'yes'

**Comment by cagnulein** [comment]
> In this case I did hit Yes the second time.

ok so that's correct what I saw. And yes, if you press yes, qz is following the new one. The no option instead should be used to ignore it, unless there is a bug on this. that's why I was asking a debug log of this scenario.
Is the "yes" action work? i mean is it following the new one then?

for the second issue i don't see any big drop of the resistance. am I missing something? I saw that you set the gears to 17 before starting, but then the resistance follow the workout applying the gear.

**Comment by marklogan17** [comment]
I did a quick test and its made me think more about my resistance drop.  The class I tested with (because I'm an idiot) was from today and had no target resistance info.  The target resistance dropped to 10 anyway.  In the XML produced by QZ the first row is:

<row duration="00:00:54" resistance="10" lower_resistance="10" upper_resistance="14" requested_peloton_resistance="25" lower_requested_peloton_resistance="25" upper_requested_peloton_resistance="35" cadence="70" lower_cadence="70" upper_cadence="80" forcespeed="0" looptimehr="10"/>

Now QZ only dropped the resistance to 20 because I had that as my minimum resistance.  I geared up to get it to 26 or 27 and it was all aligned.  So I thought maybe the only issue there is that it set the target resistance to the 'lower' instead of the 'upper' that I had configured in the app.  Wait a minute Mark, this is a newer tablet, is there any chance you're more of an idiot than you thought?  Why yes there is.  I went into the Peloton options and sure enough I have it set to 'lower'.

So I have not created a new issue because, as per normal, its me who is too stupid to configure it right :)

I'll do the test with the two peloton classes and hitting 'no' later.
